File: java/app/src/main/java/othello/ColorPrint.java
package othello;

public class ColorPrint {
    /**
     * Format string with color.
     */
    public static String getColor(String text, AnsiColor color, boolean bold) {
        // Check if the text contains a reset code already
        String coloredText = text.replace(AnsiColor.RESET.toString(), AnsiColor.RESET + color.toString());

        if (bold) {
            return color + AnsiColor.BOLD.toString() + coloredText + AnsiColor.RESET;
        }

        return color + coloredText + AnsiColor.RESET;
    }

    /**
     * Print text with specified color.
     */
    public static void printColor(String text, AnsiColor color) {
        System.out.println(getColor(text, color, false));
    }

    /**
     * Print bold text.
     */
    public static void printBold(String text) {
        printBold(text, AnsiColor.WHITE);
    }

    /**
     * Print bold text with optional color.
     */
    public static void printBold(String text, AnsiColor color) {
        System.out.println(getColor(text, color, true));
    }

    /**
     * Print error message with red color.
     */
    public static void printError(String message) {
        String[] split = splitLeadingWhitespace(message);
        System.out.println(split[0] + getColor("Error: " + split[1], AnsiColor.RED, false));
    }

    /**
     * Print warning message with yellow color.
     */
    public static void printWarn(String message) {
        String[] split = splitLeadingWhitespace(message);
        System.out.println(split[0] + getColor("Warning: " + split[1], AnsiColor.YELLOW, false));
    }

    /**
     * Split a string into the leading whitespace and the rest of the string.
     */
    private static String[] splitLeadingWhitespace(String message) {
        int indentSize = 0;
        while (indentSize < message.length() && Character.isWhitespace(message.charAt(indentSize))) {
            indentSize++;
        }
        return new String[]{message.substring(0, indentSize), message.substring(indentSize)};
    }
}

File: kotlin/app/src/main/kotlin/othello/ColorPrint.kt
package othello

enum class AnsiColor(private val code: String) {
    CYAN("\u001B[36m"),
    GREEN("\u001B[32m"),
    MAGENTA("\u001B[35m"),
    RED("\u001B[31m"),
    WHITE("\u001B[37m"),
    YELLOW("\u001B[33m"),
    BOLD("\u001B[1m"),
    RESET("\u001B[0m"),
    ;

    override fun toString(): String {
        return code
    }
}

/** Format string with colour.*/
fun getColor(text: String, color: AnsiColor, bold: Boolean = false): String {
    // Check if the text contains a reset code already
    val coloredText = text.replace(AnsiColor.RESET.toString(), "${AnsiColor.RESET}$color")

    if (bold) {
        return "$color${AnsiColor.BOLD}$coloredText${AnsiColor.RESET}"
    }

    return "$color$coloredText${AnsiColor.RESET}"
}

/** Print text with specified colour.*/
fun printColor(text: String, color: AnsiColor) {
    println(getColor(text, color))
}

/** Print bold text with optional colour.*/
fun printBold(text: String, color: AnsiColor = AnsiColor.WHITE) {
    println(getColor(text, color, bold = true))
}

/** Print error message with red colour.*/
fun printError(message: String) {
    val (indent, text) = splitLeadingWhitespace(message)
    println("$indent${getColor("Error: $text", AnsiColor.RED)}")
}

/** Print warning message with yellow colour.*/
fun printWarn(message: String) {
    val (indent, text) = splitLeadingWhitespace(message)
    println("$indent${getColor("Warning: $text", AnsiColor.YELLOW)}")
}

/** Split a string into the leading whitespace and the rest of the string.*/
fun splitLeadingWhitespace(message: String): Pair<String, String> {
    val indentSize = message.takeWhile { it.isWhitespace() }.length
    return message.substring(0, indentSize) to message.substring(indentSize)
}
